### PR TITLE
Avoid evaluating object on linear during initialization in Crank Nicolson time integrator

### LINKED ISSUE #10699: Initial residual evaluation should not evaluate objects only on linear
## Rationale
Otherwise, some side effect could show up.

## Description
This is a subtle issue revealed by one Rattlesnake heavy test. In that test, we are using `CrankNicolson` time integrator, which requires a residual evaluation on initial. That test contains an auxiliary kernel on `linear`, which is integrating another variable in time with `VariableTimeIntegrationAux`. Its corresponding auxiliary variable is coupled back in kernel evaluations. The initial residual evaluation will actually invoke that aux kernel and change the variable from its initial value. My fix in https://github.com/idaholab/moose/pull/10564 changes the behavior because the current solution is copied to the old solution before the fix but not after the fix, thus the evaluation on that aux kernel gives different values. Neither gives the correct answer though. The initial value of the variable should not be changed.

The correct way is to not update that aux variable and use the values set by the initial condition. Although currently that aux kernel is only on linear, MOOSE will call it on the initial residual evaluation.

I think the proper solution is to create one function `FEProblem::computeInitialResidual()` which is almost identical to `FEProblem::computeResidualType()` except that
1. It does not need the third argument for kernel type. The kernel type will be always `Moose::KT_NONTIME`.
2. It does not need `_nl->computeTimeDerivatives()` in the function.
3. All `EXEC_LINEAR` needs to be replaced with `EXEC_INITIAL`.

If you think this is the right move, I can submit a PR really quickly. I do need your opinions. Tag @andrsd @permcody.

## Impact
The impact should be small. It should only affect a subset of tests using Crank-Nicolson time integrator.

### PR DESCRIPTION
**Pull request by YaqiWang**
This is the solution I was talking about. It is not final, but I want to see how much damage it can cause.

Close #10699.

### CODE REVIEW

**Comment by permcody** [comment]
I'm not sure I like this solution because it's almost 100% code duplication. Can we not parameterize the execute flag instead, or just pass in a Boolean to do the one or two special things?

**Comment by andrsd** [comment]
Well, I am pretty sure I do not like this. Adding a parameter with execute flag with default value `EXEC_LINEAR` is much better solution.

**Comment by YaqiWang** [comment]
This is my new solution. Looks much nicer, right?

**Comment by YaqiWang** [comment]
The failing test in Rattlesnake is due to a bad interaction of the executioner in Rattlesnake derived from `Transient` (save-in variables are not cleared before the second residual evaluation with `_nl.computeResidual()`). The gold file is good. I have a patch ready. (All of my recent PRs and issues are actually kind of related with the cleanup of this executioner.)

The failing in Badger seems real. Looks like the failing tests are using Crank-Nicolson integrator and require some initial setup before the residual call, which is delayed for some reason. I have no access to the code, so do not know how difficult it can be fixed. I feel that the fix will be a good thing.

I have no idea why unity build failed.

Another thing is that this PR makes time integrator `init()` not necessarily done after initial time step size is calculated although should still only be done when not recovering.

This PR also fix a wrong comment on https://github.com/idaholab/moose/blob/devel/framework/src/timeintegrators/CrankNicolson.C#L42, because the later residual evaluation in that function will actually calculate `u_dot` into a nonzero vector. It does not hurt us currently because no place is using it and `compute_residual` will update it correctly later.

I think this also opens a possible integrity check for us: always doing the initial residual evaluation should not hurt applications or well-designed tests except one possible unnecessary evaluation. If it does, something must be wrong.

File: framework/src/base/FEProblemBase.C [hunk 1]
```diff
@@ -704,6 +705,14 @@ FEProblemBase::initialSetup()
     // The only user objects that should be computed here are the initial UOs
     computeUserObjects(EXEC_INITIAL, Moose::POST_AUX);
 
+    if (_initial_residual_required)
+    {
+      _nl->zeroVariablesForResidual();
```
Comment (YaqiWang) [inline framework/src/base/FEProblemBase.C]: These two lines seem redundant, I put them here to be more defensive.

File: framework/src/timeintegrators/CrankNicolson.C [hunk 2]
```diff
@@ -39,17 +40,6 @@ CrankNicolson::computeTimeDerivatives()
 void
 CrankNicolson::init()
 {
-  // make sure that time derivative contribution is zero in the first pre-solve step
-  _u_dot.zero();
```
Comment (YaqiWang) [inline framework/src/timeintegrators/CrankNicolson.C]: I can add `_u_dot.zero()` and `_du_dot_du = 0` back if you think we should.

Comment (andrsd) [reply framework/src/timeintegrators/CrankNicolson.C]: 👍

**Comment by YaqiWang** [comment]
I will need help on Badger and Unity build. This is ready to me. I need this to fix the failing heavy test in Rattlesnake. Thanks.

**Comment by YaqiWang** [comment]
Anybody can help me on Badger? I do not have access......

**Comment by andrsd** [comment]
> Anybody can help me on Badger? I do not have access......

I am looking at it now.

**Comment by YaqiWang** [comment]
As discussed with @andrsd , Badger needs to be patched. This will fail Rattlesnake in the same way as I mentioned before. If more requests come for initial residual evaluation, we can retrieve back my previous commit. But for now, this meet my requirement and minimize the modification. Thanks @andrsd 's comments.

**Review by andrsd — APPROVED** [review]

**Comment by YaqiWang** [comment]
I was wrong about Rattlesnake, the residual evaluation in that test happens before the residual evaluation in `CrankNicolson::init`, thus the save-in variables are correct. The second residual evaluation can still have the correct residual but not the save-in, which is not used later. This is complicated but it is good we have explanations for all we see ;-) Grizzly failure looks irrelevant.

**Comment by YaqiWang** [comment]
Thanks @andrsd .

**Pull request merged**